File: src/apphost.cs
#!/usr/bin/env dotnet
#:sdk Aspire.AppHost.Sdk@13.3.3
#:package Aspire.Hosting.Azure.AppContainers
#:package Aspire.Hosting.Azure.PostgreSQL
#:package Aspire.Hosting.Azure.Redis
#:package Aspire.Hosting.Docker
#:package Aspire.Hosting.Redis
#:package Aspire.Hosting.JavaScript
#:package Aspire.Hosting.Yarp
#:package Aspire.Hosting.Maui
#:package Aspire.Hosting.DevTunnels
#:project ./BingoBoard.Admin
#:project ./BingoBoard.MigrationService
#:include ./apphost/*.cs

#pragma warning disable

using Microsoft.Extensions.Hosting;

var builder = DistributedApplication.CreateBuilder(args);

var buildInfo = BuildInfo.Load(builder);
var postgresAzureLocation = builder.Configuration["Azure:PostgresLocation"];

Console.WriteLine($"Environment name: {builder.Environment.EnvironmentName}");

builder.AddAzureContainerAppEnvironment("env");

var password = builder.AddParameter("admin-password", secret: true);
var adminDomain = builder.AddParameter("admin-domain", "admin.aspireify.live");
var adminCertName = builder.AddParameter("admin-cert-name", "admin.aspireify.live-envvevso-251017190301");
var yarpDomain = builder.AddParameter("yarp-domain", "aspireify.live");
var yarpCertName = builder.AddParameter("yarp-cert-name", "aspireify.live-envvevso-251017185247");

var cache = builder.AddRedis("cache")
    .PublishAsBingoCache();

var postgres = builder.AddAzurePostgresFlexibleServer("postgres")
    .ConfigureBingoPostgres(postgresAzureLocation)
    .WithPasswordAuthentication()
    .RunAsContainer(container => container.WithLifetime(ContainerLifetime.Persistent));

var db = postgres.AddDatabase("db")
    .WithPostgresMcp();

var migrations = builder.AddProject<Projects.BingoBoard_MigrationService>("migrations")
    .WithEnvironment("Authentication__AdminPassword", password)
    .WithReference(db)
    .WaitFor(db);


var admin = builder.AddProject<Projects.BingoBoard_Admin>("boardadmin")
    .WithReference(cache)
    .WithReference(db)
    .WithReference(migrations)
    .WaitFor(cache)
    .WaitForCompletion(migrations)
    .WithBuildInfo(buildInfo)
    .WithExternalHttpEndpoints()
    .PublishAsBingoAdmin(adminDomain, adminCertName);

admin.WithImportSquaresCommand();

var frontend = builder.AddViteApp("bingoboard-dev", "./bingo-board")
    .WithBuildInfo(buildInfo, prefix: "VITE_")
    .WithReference(admin)
    .WaitFor(admin);

builder.AddYarp("bingoboard")
    .WithConfiguration(c =>
    {
        c.AddRoute("/api/version-info", admin);
        c.AddRoute("/bingohub/{**catch-all}", admin);
    })
    .PublishWithStaticFiles(frontend)
    .WaitFor(admin)
    .WithIconName("SerialPort")
    .WithExternalHttpEndpoints()
    .PublishAsBingoGateway(yarpDomain, yarpCertName)
    .WithExplicitStart();

builder.AddBingoMauiClients(admin);

builder.Build().Run();

File: src/apphost.ts
import { createBuilder, ContainerLifetime } from './.modules/aspire.js';

const builder = await createBuilder();

builder.addAzureContainerAppEnvironment("env");

// Version info (C# used reflection; in TypeScript we use env vars / package.json)
const commitSha = process.env.COMMIT_SHA ?? "dev";
const aspireVersion = process.env.ASPIRE_VERSION ?? "unknown";
const dotnetVersion = process.env.DOTNET_VERSION ?? "unknown";

// Parameters
const password = await builder.addParameter("admin-password", { secret: true });

// Redis cache
const cache = await builder.addRedis("cache");

// PostgreSQL
const postgres = await builder.addPostgres("postgres")
    .withLifetime(ContainerLifetime.Persistent);

const db = await postgres.addDatabase("db");
await db.withPostgresMcp();

// Migration service
const migrations = await builder.addProject("migrations", "./BingoBoard.MigrationService", "BingoBoard.MigrationService")
    .withEnvironmentParameter("Authentication__AdminPassword", password)
    .withReference(db)
    .waitFor(db);

// Admin / backend project
// TODO: publishAsAzureContainerApp not available on ProjectResource.
// See: https://github.com/dotnet/aspire/issues/15142, https://github.com/dotnet/aspire/issues/15152
const admin = await builder.addProject("boardadmin", "./BingoBoard.Admin", "https");
await admin.withReference(cache);
await admin.withReference(db);
await admin.withServiceReference(migrations);
await admin.waitFor(cache);
await admin.waitForCompletion(migrations);
await admin.withEnvironment("COMMIT_SHA", commitSha);
await admin.withEnvironment("ASPIRE_VERSION", aspireVersion);
await admin.withExternalHttpEndpoints();

// Vue.js frontend (Vite)
const frontend = await builder.addViteApp("bingoboard-dev", "./bingo-board")
    .withEnvironment("VITE_COMMIT_SHA", commitSha)
    .withEnvironment("VITE_DOTNET_VERSION", dotnetVersion)
    .withEnvironment("VITE_ASPIRE_VERSION", aspireVersion)
    .withServiceReference(admin)
    .waitFor(admin);

// YARP reverse proxy
// TODO: YARP route configuration blocked by codegen bug — addCluster maps to wrong capability.
const yarp = await builder.addYarp("bingoboard");
await yarp.publishWithStaticFiles(frontend);
await yarp.waitFor(admin);
await yarp.withIconName("SerialPort");
await yarp.withExternalHttpEndpoints();
// TODO: publishAsAzureContainerApp blocked. See: https://github.com/dotnet/aspire/issues/15142
await yarp.withExplicitStart();

// MAUI section (conditional on launch profile)
const launchProfile = process.env.DOTNET_LAUNCH_PROFILE;
if (launchProfile && launchProfile.toLowerCase() === "maui") {
    const adminEndpoint = await admin.getEndpoint("https");

    const publicDevTunnel = await builder.addDevTunnel("devtunnel-public", { allowAnonymous: true });
    await publicDevTunnel.withTunnelReference(adminEndpoint);

    const mauiapp = await builder.addMauiProject(
        "mauiapp",
        "BingoBoard.MauiHybrid/BingoBoard.MauiHybrid.csproj"
    );

    // iOS simulator
    // NOTE: C# used .WithReference(admin, publicDevTunnel) which is [AspireExportIgnore] in polyglot.
    // Tunnel reference is set up separately above; MAUI devices reference admin directly.
    await mauiapp.addiOSSimulator("mauiapp-ios")
        .excludeFromManifest()
        .withOtlpDevTunnel()
        .withServiceReference(admin);

    // Android emulator
    const android = await mauiapp.addAndroidEmulator("mauiapp-android");
    await android.excludeFromManifest();
    await android.withParentRelationship(mauiapp);
    await android.withOtlpDevTunnel();
    await android.withServiceReference(admin);

    // Mac Catalyst desktop
    await mauiapp.addMacCatalystDevice("mauiapp-maccatalyst")
        .excludeFromManifest()
        .withServiceReference(admin);

    // Windows desktop
    await mauiapp.addWindowsDevice("mauiapp-windows")
        .excludeFromManifest()
        .withServiceReference(admin);
}

await builder.build().run();
